website/docs/ref/react.doc.js

```diff
@@ -285,10 +307,105 @@ function loadAsync<Config>(
+  ## Higher-Order-Components II: React Class Decorators
 
-<AsyncGreeter />;
+  **NOTE**: Advanced and Experimental
+  This uses undocumented types such as $Diff which may get changed without a warning.
+  Use at your own risk.
+
+  Higher-Order-Components are often also functions that accept a ReactClass,
+  wrap it in another ReactClass and return the new ReactClass.
+
+  In such cases, the newly created ReactClass will usually accept the same `props`
+  as the original Component, but it may require a sub-set of the original props,
+  as it my inject some on it's own.
+
+  Here we will see a Class Decorator that will automatically inject the current Date
+  into a component.
+*/
+
+type TimelyProps = {
+  date: Date,
+  name: string,
+  excited: boolean
+};
+
+class Timely extends React.Component<void, TimelyProps, void> {
+  render() {
+    const hours = this.props.date.getHours();
+    const timeOfDay = hours > 17 ? 'Evening' : hours > 12 ? 'Afternoon' : 'Morning';
+
+    return (
+      <div>
+        Good {timeOfDay} {this.props.name} {this.props.excited ? '!' : ''}
+      </div>
+    );
+  }
+}
+
+// $ExpectError
+<Timely />; // Timely didn't get the required props
+
+// $ExpectError
+<Timely name='John' />; // Timely didn't get the required props
+
+// $ExpectError
+<Timely name='John' excited={true} />; // Timely didn't get the required props
+
+
+const x = <Timely date={new Date()} name='John' excited={true} />
+// Only way to satisfy all prop requirements.
+type Injected = {date: Date}
+declare function injectDate<DefaultProps, Props, State, C: React$Component<DefaultProps, Props, State>>(
+  Komponent: Class<C>
+): Class<React$Component<void, $Diff<Props, {date: Date}>, void>>;
+
+const Timeless = injectDate(Timely);
+// $ExpectError
+<Timeless />; // props not satisfied.
+
+// $ExpectError
+<Timeless excited={true} />; // props not satisfied.
+
+// $ExpectError
+<Timeless name='Sally' />; // props not satisfied.
+
+// $ExpectError
+<Timeless name={1234} excited={true} />; // name must be a string
+
+
+<Timeless name='Sally' excited={true} />;
+
+// $ExpectError
+<Timeless name='Sally' excited={true} date={1234} /> // date must still be a Date object
+
+/*
+  If you want to support stateless functions with your HOCs, the type definition gets
+  slightly more complicated. The type above will need to be written like this to support
+  stateless functions:
+*/
+
+// D = DefaultProps, P: Props, S: State;
+type InjectDateToClass = <D, P, C: React$Component<D, P, any>>(
+  Komponent: Class<C>
+) => Class<React$Component<D, $Diff<P, Injected>, any>>;
+
+type InjectDateStateless = <P>(component: (props: P) => any) => Class<React$Component<void, $Diff<P, Injected>, any>>;
+
+declare var injectDateStateless: InjectDateStateless & InjectDateToClass;
+
+
+/*
+  Here, we essentially overloaded the definition for the HOC with the types for stateless functions.
+  This is neccessary as statelss component don't have defaultProps, while classes do.
```

Comment:
the .propTypes don't matter. But it is possible to adapt the types for default Props. I chose to leave it out for simplicity as I think most developers don't use that features with stateless functions. I can re-visit this and see if I can make it work.